aten/src/ATen/native/cuda/Nonzero.cu

```diff
@@ -137,4 +165,132 @@ Tensor nonzero_cuda(const Tensor& self){
   Tensor out = at::detail::empty_cuda({0}, self.options().dtype(kLong));
   return at::native::nonzero_out_cuda(self, out);
 }
+
+template <typename scalar_t>
+void nonzero_static_cuda_out_impl(
+    const Tensor& self,
+    int size,
+    int64_t fill_value,
+    Tensor& out) {
+  Tensor self_ = self.contiguous();
+  const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  // compute number of nonzero elements
+  size_t temp_storage_bytes = 0;
+  auto& allocator = *c10::cuda::CUDACachingAllocator::get();
+  auto num_nonzeros = allocator.allocate(sizeof(int));
+  cub::TransformInputIterator<bool, NonZeroOp<scalar_t>, const scalar_t*> itr(
+      self_.const_data_ptr<scalar_t>(), NonZeroOp<scalar_t>());
+  // expected output size is num_nonzeros x ndim
+  // we are producing output with size {num_nonzeros, ndim} and strides {1,
+  // num_nonzeros} (that is, transposed ndim x num_nonzeros output) we are able
+  // to directly use passed output with this size and strides, and we can also
+  // (per contract) resize passed output with incorrect sizes anyway we want.
+  // However, out with correct sizes and incorrect strides will have to be
+  // copied to from the intermediate we've produced.
+  bool need_to_copy = out.dim() == 2 && out.sizes()[0] == size &&
+      out.sizes()[1] == self.dim() && !out.t().is_contiguous();
+  at::Tensor out_temp = need_to_copy
+      ? Tensor(at::detail::empty_cuda({self.dim(), size}, out.options()))
+      : out.resize_({self.dim(), size});
+  // Scalars are expected to produce output of size (1,0), so we can't write to
+  // it
+  if (size > 0 && self.dim() > 0) {
+    cub::CountingInputIterator<int64_t> counting_itr(0);
+    temp_storage_bytes = 0;
+    cub::DeviceSelect::Flagged(
+        nullptr,
+        temp_storage_bytes,
+        counting_itr,
+        itr,
+        out_temp.mutable_data_ptr<int64_t>(),
+        (int*)num_nonzeros.get(),
+        std::min(size, (int)self_.numel()),
+        stream);
+    auto temp_storage = allocator.allocate(temp_storage_bytes);
+    cub::DeviceSelect::Flagged(
+        temp_storage.get(),
+        temp_storage_bytes,
+        counting_itr,
+        itr,
+        out_temp.mutable_data_ptr<int64_t>(),
+        (int*)num_nonzeros.get(),
+        std::min(size, (int)self_.numel()),
```

Comment:
Hm, so to make sure that all elements are inspected I'll need to pass `size` that's equal to or greater than self_.numel() right? So if e.g. I know that there are only N nonzero elements in my tensor I can't make use of this information, and would still need to allocate full size output in this case, which oftentimes is prohibitively expensive (nonzero was implemented like this at first, but people complained about it, so we changed the implementation to first count nonzeros and then allocate output). 
Don't know how to do the right thing with cub APIs though.

Reply:
> So if e.g. I know that there are only N nonzero elements in my tensor I can't make use of this information, and would still need to allocate full size output in this case, which oftentimes is prohibitively expensive (nonzero was implemented like this at first, but people complained about it, so we changed the implementation to first count nonzeros and then allocate output).

Right. Thank you for pointing this out. size < input.numel() is a case I did not consider I can boil down your concern into this example:

```
In [4]: torch.nonzero_static(torch.tensor([1,0,1,0]),  size=2, fill_value=-1)
Out[4]: 
tensor([[0],
        [2]])

In [5]: torch.nonzero_static(torch.tensor([1,0,1,0], device="cuda"),  size=2, fill_value=-1)
Out[5]: 
tensor([[ 0],
        [-1]], device='cuda:0')
```

As far as how to handle this in a reasonable way, I am still thinking. Doing a scan for number of nonzeros up to a particular index (rather than a reduce), followed by a binary search for "size", would allow us to specify num_items correctly. But I want to think more about whether there is a better way.